MongoDB 3.6 Support

rozza opened the pull request:
Updated the Mongo Async Java Driver to 3.6.0 and updated
the API to match the new API.  This includes support for
Client Sessions, updates to List Databases and Aggregation.

JAVARS-41

jyemin commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11)







---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 202 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5z3yoYIjFoA1TrWJT)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L202)):</sup>
This is going TDocument as well if you lgtm my CR.

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 224 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zD6FVECXSfji7vUo)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L224)):</sup>
This one too

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 341 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zJ0lLjXiZ9c-4saX)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L341)):</sup>
Remove all occurrences in javadoc of: This can be of any type for which a {@code Codec} is registered
As it's no longer true.

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 480 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zaXUKMgkyftWg4Lj)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L480)):</sup>
I've been debating whether this should go to BsonDocument, even though it's awkward to use.  What do you think?

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoDatabase.java, line 110 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zujUq9z_Ce26M7Q-)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoDatabase.java#L110)):</sup>
Same here: BsonDocument?  Rather than privilege the Document class?

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 21 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-SLLOaGyAgfFwuCg)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L21)):</sup>
Just a thought: No @since tags in this project?  I just noticed it for the first time.

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 22 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-adKk1l5wg44HiL8)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L22)):</sup>
This is ugly, but I don't have any better ideas.

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 23 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-NYyeOOp41SFF7oq)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L23)):</sup>
Requires a Javadoc comment, no?  Is checkstyle not enforcing this?

---


<!-- Sent from Reviewable.io -->

rozza commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11)







---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 341 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zJ0lLjXiZ9c-4saX)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L341)):</sup>
Done.

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java, line 480 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj5zaXUKMgkyftWg4Lj)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/MongoCollection.java#L480)):</sup>
Document won out I think

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 21 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-SLLOaGyAgfFwuCg)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L21)):</sup>
The ones that were there were all @since 3.0.  I have updated to @since 1.0 
Anyway to enforce via checkstyle?

---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 23 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-NYyeOOp41SFF7oq)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L23)):</sup>
Checkstyle doesn't enforce it  - but have added

---


<!-- Sent from Reviewable.io -->

jyemin commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11)







---

<sup>**[driver/src/main/com/mongodb/reactivestreams/client/Success.java, line 23 \[r1\]](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11#-Jj6-SLLOaGyAgfFwuCg)** ([raw file](https://github.com/rozza/mongo-java-driver-reactivestreams/blob/66528c4512f8ec4b184f248d601bf9419c81ca7e/driver/src/main/com/mongodb/reactivestreams/client/Success.java#L23)):</sup>
Perhaps if we write a custom module, but not worth it right now.

---


<!-- Sent from Reviewable.io -->

jyemin commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/rozza/mongo-java-driver-reactivestreams/11)



---

LGTM

---



<!-- Sent from Reviewable.io -->

rozza commented:
Rebased and merged

rozza committed:
MongoDB 3.6 Support

Updated the Mongo Async Java Driver to 3.6.0 and updated
the API to match the new API.  This includes support for
Client Sessions, updates to List Databases and Aggregation.

JAVARS-41

jyemin commented on .travis.yml:
```diff
@@ -21,8 +18,26 @@ branches:
   global:
-    - TERM=dumb
+    - MONGODB_FILE_NAME=mongodb-linux-x86_64-enterprise-ubuntu1404
+    - MONGODB=3.4.7
```
Can't use 3.6.0 yet?

rozza replied:
Will try..

jyemin commented on build.gradle:
```diff
@@ -99,7 +99,8 @@ configure(subprojects.findAll { it.name != 'util' }) {
         options.version = true
-        options.links 'http://docs.oracle.com/javase/7/docs/api/'
+        options.links 'http://docs.oracle.com/javase/8/docs/api/'
+        options.links 'http://mongodb.github.io/mongo-java-driver/3.6/javadoc/'
```
Good catch.

jyemin commented on driver/src/main/com/mongodb/reactivestreams/client/gridfs/GridFSBucket.java:
```diff
@@ -232,51 +375,98 @@
+     *
+     * @param filename the name of the file to be downloaded
+     * @return the stream
+     */
```
@since 1.7

rozza replied:
This just got moved to keep all the same named methods inline.

jyemin replied:
Acknowledged

jyemin commented on driver/src/main/com/mongodb/reactivestreams/client/gridfs/GridFSBucket.java:
```diff
@@ -232,51 +375,98 @@
+     * @param filename the name of the file to be downloaded
+     * @param options  the download options
+     * @return the stream
+     */
```
@since 1.7

rozza replied:
This just got moved to keep all the same named methods inline.

jyemin replied:
Acknowledged

jyemin commented on driver/src/main/com/mongodb/reactivestreams/client/gridfs/GridFSBucket.java:
```diff
@@ -232,51 +375,98 @@
     /**
-     * Downloads the contents of the stored file specified by {@code id} and writes the contents to the {@code destination}
-     * AsyncOutputStream.
+     * Opens a AsyncInputStream from which the application can read the contents of the stored file specified by {@code id}.
```
Some comments say "Opens a AsyncInputStream", others say "Opens a Stream".  Let's make them consistent.

rozza replied:
Done.

jyemin commented on driver/src/main/com/mongodb/reactivestreams/client/internal/MongoClientImpl.java:
```diff
@@ -73,4 +82,24 @@ public MongoClientSettings getSettings() {
+    }
+
+    @Override
+    public Publisher<ClientSession> startSession(final ClientSessionOptions options) {
```
Add a test for this.

rozza replied:
Good catch. Done.

jyemin replied:
IntelliJ caught it with its unused method warning.

jyemin reviewed (changes requested):
Glad you did this all as one review.  Looks great, just a few nits.

rozza committed:
CodeReview updates
